examples/src/demo_client.rs

```diff
@@ -0,0 +1,154 @@
+
+    let rounds_response = response.unwrap().into_inner();
+    let oldest_round = rounds_response.oldest_round;
+    let newest_round = rounds_response.newest_round;
+    let round = oldest_round + 1;
+    println!("\n2) Find collections from earliest round and continue to add collections until gas limit is hit\n");
+    println!("\n---- Use NodeReadCausal endpoint ----\n");
+
+    let mut collection_ids: Vec<CertificateDigest> = vec![];
+    while round < newest_round && collection_ids.len() < gas_limit {
+        let request = tonic::Request::new(NodeReadCausalRequest {
+            public_key: Some(PublicKey {
+                bytes: public_key.clone(),
+            }),
+            round,
+        });
+
+        println!("NodeReadCausalRequest={:?}\n", request);
+
+        let response = client.node_read_causal(request).await;
+
+        println!("NodeReadCausalResponse={:?}\n", response);
+
+        let node_read_causal_response = response.unwrap().into_inner();
+
+        if collection_ids.len() + node_read_causal_response.collection_ids.len() <= gas_limit {
+            collection_ids.extend(node_read_causal_response.collection_ids);
+        } else {
+            println!("Reached gas limit of {gas_limit}, stopping search for more collections\n");
+            break;
+        }
+    }
+
+    println!(
+        "Proposing block with {} collections!\n",
+        collection_ids.len()
+    );
+
+    println!(
+        "\n******************************** Validator Service ********************************\n"
+    );
+    let mut client = ValidatorClient::connect("http://127.0.0.1:8000").await?;
+
+    println!("\n3) Find all causal collections from the collections found.\n");
+    println!("\n---- Use ReadCausal endpoint ----\n");
+    let node_read_causal_cids = collection_ids.clone();
+    for collection_id in node_read_causal_cids {
+        let request = tonic::Request::new(ReadCausalRequest {
+            collection_id: Some(collection_id),
+        });
+
+        println!("ReadCausalRequest={:?}\n", request);
+
+        let response = client.read_causal(request).await;
+
+        println!("ReadCausalResponse={:?}\n", response);
+
+        let read_causal_response = response.unwrap().into_inner();
+
+        collection_ids.extend(read_causal_response.collection_ids);
+    }
+
+    println!("\n4) Obtain the data payload from collections found.\n");
+    println!("\n---- Use GetCollections endpoint ----\n");
+    let request = tonic::Request::new(GetCollectionsRequest {
+        collection_ids: collection_ids.clone(),
+    });
+
+    println!("GetCollectionsRequest={:?}\n", request);
+
+    let response = client.get_collections(request).await;
+
+    println!("GetCollectionsResponse={:?}\n", response);
+
+    let get_collection_response = response.unwrap().into_inner();
+
+    // TODO: This doesn't work yet, figure out why workers are crashing.
+    println!("Found {} batches", get_collection_response.result.len());
```

Comment:
This still does not work, but not because of workers crashing:
https://gist.github.com/huitseeker/8ced40e2e8ebf14e5d1aa73b3fd2bb0f